#! /bin/sh /usr/share/dpatch/dpatch-run
## snmp_memleak.dpatch by Florian Forster <octo@noris.net>
##
## DP: Fix a possible memory leak in the snmp plugin: The SNMP result object
## DP: was not freed when `csnmp_instance_list_add' failed.
## DP: (This is upstream Git commit 0109e3c1f3a515ed716ddbdc261e0ed2e3f8e640)

@DPATCH@

diff a/src/snmp.c b/src/snmp.c
--- a/src/snmp.c
+++ b/src/snmp.c
@@ -1120,10 +1120,6 @@ static int csnmp_read_table (host_definition_t *host, data_definition_t *data)
     vb = res->variables;
     if (vb == NULL)
     {
-      if (res != NULL)
-	snmp_free_pdu (res);
-      res = NULL;
-
       status = -1;
       break;
     }
@@ -1132,10 +1128,7 @@ static int csnmp_read_table (host_definition_t *host, data_definition_t *data)
      * subtree */
     if (csnmp_check_res_left_subtree (host, data, res) != 0)
     {
-      if (res != NULL)
-	snmp_free_pdu (res);
-      res = NULL;
-
+      status = 0;
       break;
     }
 
@@ -1157,11 +1150,7 @@ static int csnmp_read_table (host_definition_t *host, data_definition_t *data)
 	  (vb != NULL) && (vb->next_variable != NULL);
 	  vb = vb->next_variable)
 	/* do nothing */;
-      if (vb == NULL)
-      {
-	status = -1;
-	break;
-      }
+      assert (vb != NULL);
 
       /* Copy OID to oid_list[data->values_len] */
       memcpy (oid_list[data->values_len].oid, vb->name,
@@ -1224,6 +1213,10 @@ static int csnmp_read_table (host_definition_t *host, data_definition_t *data)
     res = NULL;
   } /* while (status == 0) */
 
+  if (res != NULL)
+    snmp_free_pdu (res);
+  res = NULL;
+
   if (status == 0)
     csnmp_dispatch_table (host, data, instance_list, value_table);
 
